kernel-rs/src/kernel.rs

```diff
@@ -30,6 +30,9 @@ pub fn kernel() -> &'static Kernel {
     unsafe { &KERNEL }
 }
 
+/// # Safety
+///
+/// `CurrentProc`'s existence is only valid while it's `inner` is stored in `cpus`'s `Cpu`.
```

Comment:
`CurrentProc`의 `inner`가 유효한 포인터인지 결정하는 데 `inner`가 `Cpu`에 저장되어 있느냐 아니냐가 영향을 주지는 않습니다. `Cpu`가 가지고 있는 것이 그냥 raw pointer이므로 `Cpu`가 가지고 있는 포인터더라도 유효하지 않은 포인터일 수 있고, `Cpu`에 저장되어 있지 않은 포인터여도 유효할 수 있습니다.

`Cpu`가 nonnull `Proc` 포인터를 가지고 있다면 `Kernel`이 살아 있는 동안 그 `Proc` 포인터가 항상 유효하다는 것이 invariant가 되어야 할 것 같습니다. 근데 이 invariant를 보장할 수가 없는 것이, `Kernel`이 가지고 있는 `Cpu`가 가지고 있는 포인터는 `Kernel` 내부의 지점을 가리키므로 `Kernel` 자체가 move되면 포인터가 더 이상 유효하지 않습니다.

이 문제를 해결하려면 `Kernel`이 가지고 있는 `Proc`의 배열이 pin되어야 할 것 같은데, 그 대신 다음과 같이 해결할 수도 있을 것 같습니다: `Cpu`가 `Proc`의 포인터 대신 `Kernel` 안의 `Proc` 배열에 대한 index를 가지고 있고, `current_proc`에서는 해당 index에 접근해 `&Proc`을 얻는 것입니다. 이 방법을 사용하면 배열 접근이 추가되므로 성능상 손해이기는 하지만, `current_proc`이 trap에 걸렸을 때 처음에만 한 번 호출한 뒤 만든 `CurrentProc`을 계속 전달하여 사용하므로, 큰 손해는 아닐 것 같습니다. 실질적으로 동작에 변화가 있는 것도 아닙니다. 이 방법을 사용하면 memory safety가 pin 없이도 쉽게 보장될 것 같은데 어떻게 생각하시나요? @jeehoonkang 교수님 의견도 궁금합니다.

Reply:
```diff
@@ -977,24 +971,24 @@ const INITCODE: [u8; 52] = [
 ///    via swtch back to the scheduler.
 pub unsafe fn scheduler() -> ! {
     let mut c = kernel().mycpu();
-    unsafe { (*c).proc = ptr::null_mut() };
+    unsafe { (*c).proc = None };
     loop {
         // Avoid deadlock by ensuring that devices can interrupt.
         unsafe { intr_on() };

-        for p in &kernel().procs.process_pool {
+        for (i, p) in kernel().procs.process_pool.iter().enumerate() {
             let guard = p.lock();
             if guard.deref_info().state == Procstate::RUNNABLE {
                 // Switch to chosen process.  It is the process's job
                 // to release its lock and then reacquire it
                 // before jumping back to us.
                 guard.deref_mut_info().state = Procstate::RUNNING;
-                unsafe { (*c).proc = p as *const _ };
+                unsafe { (*c).proc = Some(i) };
                 unsafe { swtch(&mut (*c).context, &mut guard.deref_mut_data().context) };

                 // Process is done running for now.
                 // It should have changed its p->state before coming back.
-                unsafe { (*c).proc = ptr::null_mut() }
+                unsafe { (*c).proc = None }
             }
         }
     }
@@ -1024,10 +1018,9 @@ impl Kernel {
         let c = self.mycpu();
         let proc = unsafe { (*c).proc };
         unsafe { pop_off() };
-        if proc.is_null() {
-            return None;
-        }
-        // This is safe because p is non-null and current Cpu's proc.
-        Some(unsafe { CurrentProc::from_raw(proc) })
+        proc.map(|i| {
+            let inner = &self.procs.process_pool[i];
+            CurrentProc { inner }
+        })
     }
 }
```

이런 느낌으로요.

Reply:
- 의견: 잘 정리해주셔서 감사합니다. @Medowhill 님의 모든 논점에 동의합니다. (근데 아이디는 어떻게 읽는건가요?)

  다만 저는 pin을 하는 solution도 괜찮아보이긴 합니다. 어차피 kernel 전체가 pin되어야 함은 피할 수 없을 것 같아서요.
  @travis1829 님이 pin 관련 작업하고 있기도 합니다.

- 제안: 저는 (1) 잠정적으로 일단 지금 큰 변경 없이 merge하고, (2) @travis1829 님이 kernel을 pin할 때 여기도 적절히 바꾸는걸 제안합니다.

Reply:
어차피 `Kernel` 전체가 pin되어야 한다면 교수님께서 제안해 주신 내용이 올바른 것 같습니다. 안 그래도 지금 #394 를 보니 `Kernel`의 `bcache`도 pin되어야 하네요.

> (근데 아이디는 어떻게 읽는건가요?)

여기서 아이디가 무엇인가요..?

Reply:
@anemoneflower "`Cpu`가 nonnull `Proc` 포인터를 가지고 있다면 `Kernel`이 살아 있는 동안 그 `Proc` 포인터가 항상 유효하다"는 내용을 invariant로 추가해 주시면 될 것 같습니다.

Reply:
> 여기서 아이디가 무엇인가요..?
@Medowhill 홍재민님의 아이디인 "Medowhil"을 말씀하시는 것 같습니다.

Reply:
"메도우힐"이라고 읽습니다. 모바일 게임 영웅서기의 지명 "매도우 힐"에서 유래했습니다.